Framework/Parallel/src/IO/Chunker.cpp

```diff
@@ -0,0 +1,220 @@
+ * *not* guarenteed to be the same for all banks. */
+std::vector<Chunker::LoadRange> Chunker::makeLoadRanges() const {
+  // Find our partition.
+  size_t partitionIndex = 0;
+  int firstRankSharingOurPartition = 0;
+  for (; partitionIndex < m_partitioning.size(); ++partitionIndex) {
+    const int ranksInPartition = m_partitioning[partitionIndex].first;
+    if (firstRankSharingOurPartition + ranksInPartition > m_rank)
+      break;
+    firstRankSharingOurPartition += ranksInPartition;
+  }
+  const auto ranksSharingOurPartition = m_partitioning[partitionIndex].first;
+  const auto &ourBanks = m_partitioning[partitionIndex].second;
+
+  // Assign all chunks from all banks in this partition to ranks in round-robin
+  // manner.
+  int64_t chunk = 0;
+  std::vector<LoadRange> ranges;
+  for (const auto bank : ourBanks) {
+    size_t current = 0;
+    while (current < m_bankSizes[bank]) {
+      if (chunk % ranksSharingOurPartition ==
+          (m_rank - firstRankSharingOurPartition)) {
+        size_t count =
+            std::min(current + m_chunkSize, m_bankSizes[bank]) - current;
+        ranges.push_back(LoadRange{bank, current, count});
+      }
+      current += m_chunkSize;
+      chunk++;
+    }
+  }
+
+  // Compute maximum chunk count (on any rank).
+  int64_t maxChunkCount = 0;
+  for (const auto &partition : m_partitioning) {
+    size_t chunksInPartition = 0;
+    for (const auto bank : partition.second)
+      chunksInPartition += m_chunkCounts[bank];
+    int ranksInPartition = partition.first;
+    int64_t maxChunkCountInPartition =
+        (chunksInPartition + ranksInPartition - 1) / ranksInPartition;
+    maxChunkCount = std::max(maxChunkCount, maxChunkCountInPartition);
+  }
+  ranges.resize(maxChunkCount);
+
+  return ranges;
+}
+
+/** Returns a vector of partitions of work (sizes) for given number of workers.
+ *
+ * The `sizes` argument defines the amount of work for a series of tasks. Here
+ * the task would be loading/processing a certain number of chunks from a file.
+ * The `workers` argument gives the number of workers to be used to process all
+ * tasks. The returned partitioning fulfils the following:
+ * - A task may be shared among workers. If workers share a task, all tasks they
+ *   are working on are shared among that group of workers.
+ * - Groups of workers and distribution of tasks to groups tries to balance
+ *   work, such that each worker has a roughly similar amount of work.
+ * Note that this method is public and static to allow for testing. */
+std::vector<std::pair<int, std::vector<size_t>>>
+Chunker::makeBalancedPartitioning(const int workers,
+                                  const std::vector<size_t> &sizes) {
+  const auto totalSize =
+      std::accumulate(sizes.begin(), sizes.end(), static_cast<size_t>(0));
+  // Indexed size vector such that we can sort it but still know original index.
+  // Elements are <size, original index, done flag>
```

Comment:
can you also put this comment in the buildPartition method doxygen. I think that will solve my aforementioned comment.